r_code/panel_estimators.R

```diff
@@ -0,0 +1,750 @@
+    !!as.name(time_var), !!as.name(outcome_var)
+  ) %>%
+    dplyr::select(-tidyselect::all_of(id_var)) %>%
+    as.matrix() %>%
+    t()
+  # Split the treated data, taking only the pretreatment periods for training.
+  split_treat_data_pre <- treat_data %>%
+    dplyr::filter(!!as.name(time_var) < Treatment_Period) %>%
+    dplyr::select(tidyselect::all_of(c(id_var, outcome_var))) %>%
+    dplyr::group_by(!!as.name(id_var)) %>%
+    dplyr::group_split(.keep = F) %>%
+    lapply(., function(x) x %>% dplyr::pull())
+
+  # For each treated unit, find when it was treated.
+  list_of_treat_times <- treat_data %>%
+    split(.[[id_var]]) %>%
+    lapply(., function(x) {
+      x %>%
+        dplyr::select(Treatment_Period) %>%
+        dplyr::slice(1) %>%
+        dplyr::pull()
+    })
+  # Parallel computation of the SCM estimates, unit by unit.
+  list_inputed_y <- furrr::future_map2(
+    .x = split_treat_data_pre,
+    .y = list_of_treat_times,
+    .f = ~ .SCMImpute(
+      treat_data = .x,
+      control_mat = control_matrix,
+      treat_time = .y
+    )
+  )
+  # Combine the list of estimated series into a dataframe, compute effects.
+  flex_scm_series <- Map(
+    dplyr::bind_cols, treat_data %>% split(.[[id_var]]),
+    lapply(list_inputed_y, function(x) {
+      return(tibble::tibble("point.pred" = x))
+    })
+  ) %>%
+    dplyr::bind_rows() %>%
+    dplyr::rename(response = !!as.name(outcome_var)) %>%
+    dplyr::mutate(
+      point.effect = response - point.pred,
+      pct.effect = (response / point.pred) - 1
+    )
+  # Compute the counterfactual effects if relevant.
+  if (!is.null(counterfac_var)) {
+    flex_scm_series <- flex_scm_series %>%
+      dplyr::mutate(
+        cf_point.effect = (response - !!as.name(counterfac_var)),
+        cf_pct.effect = (response / !!as.name(counterfac_var)) - 1
+      )
+  }
+
+  # Reorder the columns for consistency across methods.
+  flex_scm_series <- flex_scm_series %>%
+    dplyr::select(
+      tidyselect::all_of(c(
+        time_var, id_var, "response",
+        "point.pred", "point.effect",
+        "pct.effect", "Treatment_Period"
+      )),
+      dplyr::everything()
+    )
+
+  return(flex_scm_series)
+}
+
+#' Compute the unconstrained SCM weights for a given treated units.
+#'
+#' @param treat_data Treated series for a unit, over all time.
+#' @param control_mat Matrix of control series, over all time.
+#' @param treat_time Time when the treatment is assigned, splits pre and post.
+#'
+#' @return Predicted series of the outcome for the treated unit (over all time)
+#'    based on Elastic Net Penalty trained on the pre period data.
+.SCMImpute <- function(treat_data, control_mat, treat_time) {
+  # Define the control data matrix for the training period.
+  control_pre <- control_mat[seq_len(treat_time - 1), ]
+  # Fit a speedy Cross-Model Selection and Averaging Grid for ENP.
+  fit_enp <- bigstatsr::big_spLinReg(bigstatsr::as_FBM(control_pre),
```

Comment:
Does this always succeed?

Reply:
I believe it throws it's own errors in strange cases, which are probably more useful than me wrapping them?

Reply:
The traceback makes it pretty clear what broke, too. Just tested some cases.